packages/apputils/src/widgettracker.ts

```diff
@@ -0,0 +1,352 @@
+   * restored from a serialized state.
+   *
+   * #### Notes
+   * Most client code will not need to use this, since they can wait
+   * for the whole application to restore. However, if an extension
+   * wants to perform actions during the application restoration, but
+   * after the restoration of another widget tracker, they can use
+   * this promise.
+   */
+  readonly restored: Promise<void>;
+
+  /**
+   * A signal emitted when a widget is updated.
+   */
+  readonly widgetUpdated: ISignal<this, T>;
+
+  /**
+   * Find the first instance in the tracker that satisfies a filter function.
+   *
+   * @param - fn The filter function to call on each instance.
+   *
+   * #### Notes
+   * If nothing is found, the value returned is `undefined`.
+   */
+  find(fn: (obj: T) => boolean): T | undefined;
+
+  /**
+   * Iterate through each instance in the tracker.
+   *
+   * @param fn - The function to call on each instance.
+   */
+  forEach(fn: (obj: T) => void): void;
+
+  /**
+   * Filter the instances in the tracker based on a predicate.
+   *
+   * @param fn - The function by which to filter.
+   */
+  filter(fn: (obj: T) => boolean): T[];
+
+  /**
+   * Check if this tracker has the specified instance.
+   *
+   * @param obj - The object whose existence is being checked.
+   */
+  has(obj: Widget): boolean;
+
+  /**
+   * Inject an instance into the widget tracker without the tracker handling
+   * its restoration lifecycle.
+   *
+   * @param obj - The instance to inject into the tracker.
+   */
+  inject(obj: T): void;
+}
+
+/**
+ * A class that keeps track of widget instances on an Application shell.
+ *
+ * @typeparam T - The type of widget being tracked. Defaults to `Widget`.
+ *
+ * #### Notes
+ * The API surface area of this concrete implementation is substantially larger
+ * than the widget tracker interface it implements. The interface is intended
+ * for export by JupyterLab plugins that create widgets and have clients who may
+ * wish to keep track of newly created widgets. This class, however, can be used
+ * internally by plugins to restore state as well.
+ */
+export class WidgetTracker<T extends Widget = Widget>
+  implements IWidgetTracker<T>, IRestorable<T> {
+  /**
+   * Create a new widget tracker.
+   *
+   * @param options - The instantiation options for a widget tracker.
+   */
+  constructor(options: WidgetTracker.IOptions) {
+    const focus = (this._focusTracker = new FocusTracker());
+    const pool = (this._pool = new RestorablePool(options));
+
+    this.namespace = options.namespace;
+
+    focus.currentChanged.connect((_, current) => {
+      if (current.newValue !== this.currentWidget) {
+        pool.current = current.newValue;
+      }
+    }, this);
+
+    pool.added.connect((_, widget) => {
+      this._widgetAdded.emit(widget);
+    }, this);
+
+    pool.currentChanged.connect((_, widget) => {
+      if (widget === null && focus.currentWidget) {
```

Comment:
This is a little subtle, as it makes a possible change to `pool.current` in the change handler for `pool.current`. It seems correct, but maybe a comment would be useful.